db/db_impl.cc

```diff
@@ -1795,7 +1796,7 @@ Status DBImpl::WriteLevel0TableForRecovery(int job_id, ColumnFamilyData* cfd,
       s = BuildTable(
           dbname_, env_, *cfd->ioptions(), mutable_cf_options, env_options_,
           cfd->table_cache(), iter.get(),
-          ScopedArenaIterator(mem->NewRangeTombstoneIterator(ro, &arena)),
+          ScopedArenaIterator(mem->NewRangeTombstoneIterator(&arena)),
```

Comment:
the optimization you suggested is on RangeDelAggregator, not MemtableIterator. i don't think it'll be much better than checking whether the map is empty, which it is when ignore_range_deletions == true due to the higher-level checks that prevent range tombstones from being inserted into the map.

regarding checking ignore_range_deletions in the low-level iterators, it's possible for MemTable::NewRangeTombstoneIterator() but not for BlockBasedTable::NewRangeTombstoneIterator() since TableCache::NewIterator() loads the table before creating an iterator, so we avoid loading a table unnecessarily by performing the check at a higher level.

for now, let me add ReadOptions back to MemTable::NewRangeTombstoneIterator() and return an empty iterator when ignore_range_deletions == true. in the future we can add it also to BlockBasedTable::NewRangeTombstoneIterator() if we decide to use its other fields, like snapshot, for optimizations.